pkg/controller/workload/dns.go

```diff
@@ -0,0 +1,261 @@
+/*
+ * Copyright The Kmesh Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package workload
+
+import (
+	"net"
+	"net/netip"
+	"sync"
+	"time"
+
+	"google.golang.org/protobuf/proto"
+
+	"kmesh.net/kmesh/api/v2/workloadapi"
+	"kmesh.net/kmesh/pkg/controller/workload/cache"
+	"kmesh.net/kmesh/pkg/dns"
+)
+
+const (
+	WorkloadDnsRefreshRate = 200 * time.Millisecond // 200ms, used for workload dns refresh rate
+)
+
+type dnsController struct {
+	workloadsChan chan []*workloadapi.Workload
+	cache         cache.WorkloadCache
+	dnsResolver   *dns.DNSResolver
+	// store the copy of pendingResolveWorkload.
+	workloadCache map[string]*pendingResolveDomain
+	// store all pending hostnames in the workloads
+	pendingHostnames map[string][]string
+	sync.RWMutex
+}
+
+// pending resolve domain info of Dual-Engine Mode,
+// workload is used for create the apiworkload
+type pendingResolveDomain struct {
+	Workloads   []*workloadapi.Workload
+	RefreshRate time.Duration
+}
+
+func NewDnsController(cache cache.WorkloadCache) (*dnsController, error) {
+	resolver, err := dns.NewDNSResolver()
+	if err != nil {
+		return nil, err
+	}
+	return &dnsController{
+		workloadsChan:    make(chan []*workloadapi.Workload),
+		cache:            cache,
+		dnsResolver:      resolver,
+		workloadCache:    make(map[string]*pendingResolveDomain),
+		pendingHostnames: make(map[string][]string),
+	}, nil
+}
+
+func (r *dnsController) Run(stopCh <-chan struct{}) {
+	go r.dnsResolver.StartDnsResolver(stopCh)
+	go r.refreshWorker(stopCh)
+	go r.processWorkloads()
+	go func() {
+		<-stopCh
+		close(r.workloadsChan)
+	}()
+}
+
+func (r *dnsController) processWorkloads() {
+	for workloads := range r.workloadsChan {
+		r.processDomains(workloads)
+	}
+}
+
+func (r *dnsController) processDomains(workloads []*workloadapi.Workload) {
+	domains := getPendingResolveDomain(workloads)
+
+	// store all pending hostnames of clusters in pendingHostnames
+	for _, workload := range workloads {
+		workloadName := workload.GetName()
+		hostname := workload.GetHostname()
+		if _, ok := r.pendingHostnames[workloadName]; !ok {
+			r.pendingHostnames[workloadName] = []string{}
+		}
+		r.pendingHostnames[workloadName] = append(r.pendingHostnames[workloadName], hostname)
```

Comment:
another problem if this function is called multiple times with same workloads, this will store duplicate hostname